File: BOJ/1753_최단경로.cpp
#include <iostream>
#include <vector>
#include <algorithm>
#include <queue>
#include <functional>
#define endl '\n'
using namespace std;

int dist[20001];
vector<pair<int, int>> adj[20001];
int n, m;
int inf = 0xfffffff;
bool visited[20001] = { false, };

void dijkstra(int src) 
{
	priority_queue<pair<int, int>, vector<pair<int,int>>, greater<pair<int,int>>> pq;
	pq.push(make_pair(dist[src], src));

	while (!pq.empty()) {
		int node = pq.top().second;
		pq.pop();
		visited[node] = true;
		for (int i = 0; i < adj[node].size(); i++) {
			if (dist[adj[node][i].first] > adj[node][i].second + dist[node]) {
				dist[adj[node][i].first] = adj[node][i].second + dist[node];
				if (!visited[adj[node][i].first]) {
					pq.push(make_pair(dist[adj[node][i].first], adj[node][i].first));
				}
			}
		}
	}
	return;
}

int main() {
	int start;
	cin >> n >> m >> start;

	for (int i = 0; i < m; i++) {
		int u, v, w;
		cin >> u >> v >> w;
		adj[u].push_back(make_pair(v,w));
	}

	for (int i = 1; i <= n; i++) dist[i] = inf;
	dist[start] = 0;
	dijkstra(start);


	for (int i = 1; i <= n; i++) {
		if (dist[i] == inf) cout << "INF" << endl;
		else cout << dist[i] << endl;
	}
	return 0;
}

File: BOJ/1753_최단경로.java
import java.io.*;
import java.util.*;

class Node {
	int name, value;
	Node(int name, int value ){
		this.name = name;
		this.value = value;
	}
}

public class Main {
	static BufferedReader br = new BufferedReader(new InputStreamReader(System.in));
	static StringTokenizer tk;
	static int v, e, start;
	static int[] dist;
	static boolean[] visited;
	static ArrayList<ArrayList<Node>> graph = new ArrayList<>();
	
	static void dijkstra(int src) {
		PriorityQueue<Node> pq = new PriorityQueue<>(new Comparator<Node>() {
			@Override
			public int compare(Node o1, Node o2) {
				if(o1.value == o2.value) return Integer.compare(o1.name, o2.name);
				return Integer.compare(o1.value, o2.value);
			}
		});
		
		pq.add(new Node(src, dist[src]));
		while(!pq.isEmpty()) {
			Node cur = pq.poll();
			visited[cur.name] = true;
			for(int i=0; i<graph.get(cur.name).size(); i++) {
				Node next = graph.get(cur.name).get(i);
				if(dist[next.name] > next.value + dist[cur.name]) {
					dist[next.name] = next.value + dist[cur.name];
					if(!visited[next.name]) {
						pq.add(new Node(next.name, dist[next.name]));
					}
				}
			}
		}
		
	}
	public static void main(String[] args) throws IOException {
		tk = new StringTokenizer(br.readLine());
		v = Integer.parseInt(tk.nextToken());
		e = Integer.parseInt(tk.nextToken());
		
		for(int i=0; i<=v; i++) graph.add(new ArrayList<>());
		
		start = Integer.parseInt(br.readLine());
		for(int i=0; i<e; i++) {
			tk = new StringTokenizer(br.readLine());
			int src, dest, w;
			src = Integer.parseInt(tk.nextToken());
			dest = Integer.parseInt(tk.nextToken());
			w = Integer.parseInt(tk.nextToken());
			
			graph.get(src).add(new Node(dest, w));
		}
		dist = new int[v+1];
		visited = new boolean[v+1];
		
		for(int i=1; i<=v; i++) dist[i] = Integer.MAX_VALUE;
		dist[start] = 0;
		
		dijkstra(start);
		for(int i=1; i<=v; i++) {
			if(dist[i] == Integer.MAX_VALUE) System.out.println("INF");
			else System.out.println(dist[i]);
		}
	}

}

